### ISSUE TITLE
Can't figure out how to properly resample decoded audio

### ISSUE THREAD

**Issue by nstafie** [meh/rust-ffmpeg#59]
Hi,

I'm trying to use ffmpeg to decode a file into s16le PCM samples, resampling to a new sample rate if necessary. So far, I've managed to get output that matches what I'd get by subprocessing ffmpeg, however, it's incomplete. 

Every frame I get from the API matches what I get subprocessing, except the API is short a few (couple of hundred) frames. Now, If I change the sample rate on the resampler, I get the same amount of samples, but they don't match anymore, which leads me to believe I'm doing something terribly wrong somewhere, because something is changing when I modify the sample rate I create the resampler with, but, if I change the output sample rate of the subprocess, I get the same amount of samples as I do with the API resampler set to any value.

To recap, whenever I change the sample rate of the subprocessed ffmpeg, the amount of samples I get varies (which is what I'd expect), but when I change the sample rate of the resampler I create, I get the same amount of samples but with different values.

I've written an example project that exhibits the issue: https://github.com/nstafie/rust-ffmpeg-test (it takes an audio file as an argument, I've included a 44100 hz flac file in the repo)

Now, I'm not sure what else I should try, I've surely missed something, so I would really appreciate some help.

Thanks!

**Comment by meh** [comment]
I don't understand something, is the only thing you're trying to do to resample and use the resampled stuff internally (for instance to play the sound) or to re-encode it and save it?

Because if the former you don't need to use filters at all, all you need is the resampler, check [this](https://github.com/meh/rust-openal/blob/master/examples/play.rs).

**Comment by meh** [comment]
Also added a `ChannelLayout::default(channels)` method so you don't need to use the sys crate too.

**Comment by nstafie** [comment]
I do want to use the resampled data internally (well, encode it with opus) but I'd also like to filter it so that I do stuff like change the volume or whatever.

I did also look at that example, copied it as closely as I could, but I still got the same result.

> Also added a ChannelLayout::default(channels) method so you don't need to use the sys crate too.

Thanks!

**Comment by meh** [comment]
I just remembered something, in ffmpeg in various functions that may output **more** frames from the input frame, they tend to accept `NULL` as input and fill the output frame with the remaining internal frames.

I need to check if any of the functions being used can do that and add `flush` methods accordingly, like `Encoder` does.

Does that seem consistent with what you're seeing?

**Comment by nstafie** [comment]
Yes! If I set the subprocess ffmpeg output to the same sample rate, it'll output the same amount of frames. It makes sense to me then that it would output as many frames as the input, and then a flush would be required to get the extra frames.

**Comment by meh** [comment]
Ok I checked around and I think I actually had already implemented that.

Basically when you use `run` it will return `Ok(None)` if there are no more internal frames, and `Ok(Some(Delay {..}))` if there are internal frames to pull out (the `Delay` will tell you how long the remaining frames are, but you don't really have to care about it).

To get those frames you have to call `next` until it returns `None`. Now I'm not sure why I called it `next` instead of `flush`, I'm going to push a commit that **renames** it to `flush` so it's consistent with the rest.

TL;DR: try and call `flush(&mut output)` until it returns `Ok(None)` if `run(&input, &mut output)` returns `Ok(Some(Delay { .. }))`.

**Comment by nstafie** [comment]
Okay! So, I've done that and it kind of works: it gives me the entirety of the song, however, it also gives me even more samples, which seems to be garbage data (sounds horribly distorted when encoded with opus).

Any ideas?

**Comment by meh** [comment]
I'll need to check more thoroughly but I don't think I'll have time until the weekend.

I suspect the part that returns the `Delay` keeps returning the same thing regardless of how many internal frames you ate, which may be the reason you're getting *more* frames now.

**Comment by nstafie** [comment]
I did some more testing, and the actual value of the `Some(Delay)` does change after every frame I get from the flush, and of course the resulting audio from those frames is correct, there are just extra frames.

I dumped the PCM produced and played the "garbage data" back, and it actually sounds like a sped up version of the song played in reverse, although it's not the whole song, just about 7 seconds of data.

I'm not sure what else to try...

**Comment by nstafie** [comment]
Figured I should update, I still haven't got it working. 

I put my project aside for a while but I'd like to give it another shot. I wrote some example C code that resembles what the Rust code does, and dumps PCM frames to stdout: https://gist.github.com/nstafie/21c06d60f047edff746af13ac38033e4, and a simpler example that doesn't filter the audio, just decodes and resamples: https://gist.github.com/nstafie/d0baa6f847ba38ce464956f547362ff0.

Compile with `gcc main.c -lavformat -lavcodec -lavutil -lavfilter -lswresample -o decoder` and run with `./decoder input-file.mp3 > output.bin`.

Any ideas?
